ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1139,49 +1142,49 @@ MessageFormatFlags getOperationFlag() {
     }
 
     @Override
-    protected void maybeProcessCallbacks() {
-      if (progressTracker.isCryptoJobRequired() && decryptCallbackResultInfo.decryptJobComplete) {
-        decryptJobMetricsTracker.onJobResultProcessingStart();
-        if (decryptCallbackResultInfo.exception != null) {
-          decryptJobMetricsTracker.incrementOperationError();
-          logger.trace("Decryption job callback invoked with exception for {} blob {} ", blobType, blobId,
-              decryptCallbackResultInfo.exception);
-          setOperationException(
-              new RouterException("Exception thrown on decrypting content for " + blobType + " blob " + blobId,
-                  decryptCallbackResultInfo.exception, RouterErrorCode.UnexpectedInternalError));
-          progressTracker.setCryptoJobFailed();
+    protected void processCallbacks() {
+      decryptJobMetricsTracker.onJobResultProcessingStart();
+      if (decryptCallbackResultInfo.exception != null) {
+        decryptJobMetricsTracker.incrementOperationError();
+        logger.trace("Decryption job callback invoked with exception for {} blob {} ", blobType, blobId,
+            decryptCallbackResultInfo.exception);
+        setOperationException(
+            new RouterException("Exception thrown on decrypting content for " + blobType + " blob " + blobId,
+                decryptCallbackResultInfo.exception, RouterErrorCode.UnexpectedInternalError));
+        progressTracker.setCryptoJobFailed();
+      } else {
+        // in case of Metadata blob, only user-metadata needs decryption if the blob is encrypted
+        if (blobType == BlobType.MetadataBlob) {
+          logger.trace("Processing stored decryption callback result for Metadata blob {}", blobId);
+          initializeDataChunks();
```

Comment:
For 100% read-after-write across thread guarantee, the variables set in `initializeDataChunks()`, `blobInfo`, etc. should probably be volatile (possible I am misinterpreting the documentation). However, I suspect that we have a few other cases where this was not done which have not caused (discoverable) issues in practice :P 

https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4.4